### PR TITLE
[webgpu] Further tweak vectorized NaN handling in binary ops

### PR DESCRIPTION
**Pull request by hujiajie**
The big picture will be:

```
  fn binaryOperation(a : vec4<f32>, b : vec4<f32>) -> vec4<f32> {
    let isNaN = false;
    {
      // op-specific code begins
      let result = ...;
      let isNaN = ...;  // optional, declare this only if the op defines
                        // its own extra rules about NaN, say mod(0, 0)
      // op-specific code ends

      return select(result, vec4<f32>(valueForNaN),
                    vec4<bool>(isNaN) | isnanVec4(a) | isnanVec4(b));
    }
  }
```

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7185)

### CODE REVIEW

File: tfjs-backend-webgpu/src/binary_op_util.ts [hunk 1]
```diff
@@ -197,7 +186,9 @@ const POW_VEC4 = `
     resultTemp.a = 1.0;
   }
   let isNaN = (a < vec4<f32>(0.0)) & (floor(b) < b);
-  return select(resultTemp, vec4<f32>(uniforms.NAN), isNaN);
+  let valueForNaN = uniforms.NAN;
+  ${CHECK_NAN_SNIPPET_VEC4}
+  return resultTemp;
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/binary_op_util.ts:191]: The original code only checks `isNan`. Here you also add `isnanVec4(a) | isnanVec4(b)`. Is it expected?

Comment (hujiajie) [reply tfjs-backend-webgpu/src/binary_op_util.ts:191]: It's almost certain `binOp(NaN, b)` or `binOp(a, NaN)` will return `valueForNaN`. I guess previously this case was just quietly covered by the `pow()` builtin.

File: tfjs-backend-webgpu/src/binary_op_webgpu.ts [hunk 2]
```diff
@@ -87,7 +87,10 @@ export class BinaryOpProgram implements WebGPUProgram {
     const dType = this.isVec4 ? 'vec4<f32>' : 'f32';
     const opFnStr = `
     fn binaryOperation(a : ${dType}, b : ${dType}) -> ${dType} {
-      ${getBinaryOpString(this.op, this.isVec4)}
+      let isNaN = false;
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/binary_op_webgpu.ts:90]: Can you add a comment to `isNaN` in the code? Like you did in the commit message. And in `MOD_VEC4` and `POW_VEC4`. You can directly use `isNaN = ...`.
```
     // Define op own extra rules about NaN, say mod(0, 0).
      var isNaN = ${this.isVec4 ?  'vec4<bool>(false)' : 'false'};

**Review by qjia7 — APPROVED** [review]

File: tfjs-backend-webgpu/src/binary_op_webgpu.ts (hunk 2 above)
Comment (hujiajie) [reply tfjs-backend-webgpu/src/binary_op_webgpu.ts:90]: I feel it's better to have a combined comment after `valueForNaN` is restructured similarly, or even later after everything calms down.

The `let` before `isNaN` in op-specific code is an intentional design. IMO, the snippet looks more self-contained that way, so I won't wonder where is the declaration of this `isNaN` identifier while reading `MOD_VEC4`.

Comment (qjia7) [reply tfjs-backend-webgpu/src/binary_op_webgpu.ts:90]: Sounds good.

**Review by qjia7 — APPROVED** [review]
Still LGTM, thanks.

**Review by xhcao — APPROVED** [review]
LGTM, thanks.

**Review by gyagp — APPROVED** [review]
LGTM

**Pull request merged**